test-e2e/test_zookeeper_backup.py

```diff
@@ -0,0 +1,407 @@
+"""
+Tests for the ZooKeeper backup/restore scripts.
+"""
+
+import logging
+import subprocess
+import textwrap
+import uuid
+from pathlib import Path
+from typing import List
+
+import pytest
+from _pytest.fixtures import SubRequest
+from cluster_helpers import wait_for_dcos_oss
+from dcos_e2e.backends import Docker
+from dcos_e2e.cluster import Cluster
+from dcos_e2e.node import Output
+from kazoo.client import KazooClient
+from kazoo.exceptions import NoNodeError
+
+
+LOGGER = logging.getLogger(__name__)
+
+# Arbitrary value written to ZooKeeper.
+FLAG = b'flag'
+
+
+def _zk_set_flag(hosts: str, ephemeral: bool = False) -> str:
+    """
+    Store the `FLAG` value in ZooKeeper in a random Znode.
+    """
+    znode = '/{}'.format(uuid.uuid4())
+    zk_client = KazooClient(hosts=hosts)
+    zk_client.start()
+    zk_client.create(znode, makepath=True, ephemeral=ephemeral)
+    zk_client.set(znode, FLAG)
+    zk_client.stop()
+    return znode
+
+
+def _zk_check_flag(hosts: str, znode: str) -> None:
+    """
+    The `FLAG` value is found in ZooKeeper at the `znode` path.
+    """
+    zk_client = KazooClient(hosts=hosts, read_only=True)
+    zk_client.start()
+    value = zk_client.get(znode)
+    zk_client.stop()
+    assert value[0] == FLAG
+
+
+def _bash(cmd: List[str]) -> None:
+    """
+    Execute `cmd` in a subprocess, log potential errors.
+    """
+    LOGGER.debug('BASH ARGS: {}'.format(' '.join(cmd)))
+    try:
+        result = subprocess.check_output(
+            args=' '.join(cmd),
+            stderr=subprocess.STDOUT,
+            shell=True,
+        )
+        LOGGER.debug(result.decode())
+    except subprocess.CalledProcessError as exc:
+        LOGGER.error(exc.output.decode())
+        raise
+
+
+BACKUP_SCRIPT = textwrap.dedent(
```

Comment:
❓ (Optional) - if this is in a separate file that you read here then that means that:

* We have an easier transition for when this is on a remote server
* Folks who edit it can use their own tools (shellcheck!) while editing this.

Reply:
:+1: